### ISSUE TITLE
Guidance on reproducing experiments in paper

### ISSUE THREAD

**Issue by bkj** [dselsam/neurosat#2]
Hi --

I see the disclaimer about how this repo doesn't include the code to reproduce the experiments in the paper, but are you able to sketch out what I'd have to do to reproduce some of those experiments?  (In particular, I'm most interested in reproducing the results in Table 2, where you show that the learned solver can be applied to SAT-encoded version of other NP problems, but also in the `SR(U(40))` experiments)

__EDIT__: More specifically, a couple of things that could help get me off the ground -- For the experiments described in Table 1, how many problem instances did you train on?  How many epochs of training? 

Thanks
Ben

**Comment by bkj** [comment]
I'm trying to reproduce the `SR(U(40))` experiments, but am having trouble getting the model to converge.  Any thoughts on why this might not be working?

```
for ty in "train" "test" ; do
    rm -rf data/$ty/sr5; mkdir -p data/$ty/sr5
        rm -rf dimacs/$ty/sr5/grp$i;
        mkdir -p dimacs/$ty/sr5/grp$i
        python python/gen_sr_dimacs.py \
            --out_dir dimacs/$ty/sr5/grp$i --n_pairs 10000 \
            --min_n 10 --max_n 40 --py_seed 1 --np_seed 1
        python python/dimacs_to_data.py dimacs/$ty/sr5/grp$i data/$ty/sr5 12000
done;

python ./python/train.py \
    --train_dir data/train/sr5 \
    --n_epochs 50 \
    --run_id 100
```

yields 

```
epoch=0 | counter=4346 | n_var=21 | cost=0.610173 | is_sat=[False]
epoch=0 | counter=4347 | n_var=28 | cost=0.785378 | is_sat=[True]
epoch=0 | counter=4348 | n_var=30 | cost=0.613450 | is_sat=[False]
epoch=0 | counter=4349 | n_var=33 | cost=0.612451 | is_sat=[False]
epoch=0 | counter=4350 | n_var=21 | cost=0.609838 | is_sat=[False]
epoch=0 | counter=4351 | n_var=18 | cost=0.784679 | is_sat=[True]
epoch=0 | counter=4352 | n_var=22 | cost=0.783188 | is_sat=[True]
epoch=0 | counter=4353 | n_var=32 | cost=0.782364 | is_sat=[True]
epoch=0 | counter=4354 | n_var=39 | cost=0.608814 | is_sat=[False]
epoch=0 | counter=4355 | n_var=19 | cost=0.607541 | is_sat=[False]
epoch=0 | counter=4356 | n_var=32 | cost=0.608002 | is_sat=[False]
```
This doesn't look like it's converged at all -- the cost is high when `is_sat=True` and low when `is_sat=False` which is consistent with the model predicting the same thing every time.   Any ideas what might be going on?

(I got that printout by modifying `train_epoch` to be:
```
counter = 0
for problem in train_problems:
    d = self.build_feed_dict(problem)
    _, logits, cost = self.sess.run([self.apply_gradients, self.logits, self.cost], feed_dict=d)
    epoch_train_cost += cost
    epoch_train_mat.update(problem.is_sat, logits > 0)
    print('epoch=%d | counter=%d | n_var=%d | cost=%f | is_sat=%s' %\
        (epoch, counter, problem.n_vars, cost, str(problem.is_sat)), file=sys.stderr)
    counter += 1
```
)

**Comment by dselsam** [comment]
FYI I added a wiki entry explaining the reproducibility situation in more detail (https://github.com/dselsam/neurosat/wiki/On-reproducing-the-experiments). This explanation is based on an email I sent to you before you posted this issue, and probably does not contain any new information for you.

> More specifically, a couple of things that could help get me off the ground -- For the experiments described in Table 1, how many problem instances did you train on? How many epochs of training?

We performed one epoch but on the order of 10 million problems. It is performing very well way before then though, and would probably work with many passes over a smaller set of problems. 

Note: you can't put 10 million files into a single directory. A crucial (omitted) optimization is fusing the SR-generation with the dimacs-to-data computations to prevent writing each individual file to the file system. This fused approach is omitted because we generated the fused data using multiple CPUs on GCE, and sent compressed versions back to a central Bucket. If I recall, the uncompressed dataset size is over 100 GB.

Note that the current unfused implementation isn't even batching at all when you sample `n` from a distribution, though this could be fixed (hackily) just by sorting the filenames in the directory to arrange by `n`.

> This doesn't look like it's converged at all

This comment doesn't make sense to me. Networks are not supposed to suffer the same cost for every single example in the dataset. In your example the network seems to be guessing `unsat` with very low confidence for every problem.

**Comment by bkj** [comment]
Ah ok -- thanks for the detailed writeup (and apologies if I asked a repetitive question -- not sure I ever got/saw your email).  It looks like the main difference is the number of examples -- I was training for multiple epochs on 10ks of examples, but you trained on 100-1000x more.   I will rerun with more examples and report back.

By it "doesn't look like it's converged at all", I meant that you expect the training loss to go down, but it was staying constant.  You'd expect training loss to go down _especially_ if the training set is too small -- but I didn't see any overfitting either.

**Comment by dselsam** [comment]
Closing due to inactivity.

**Comment by bywbilly** [comment]
> FYI I added a wiki entry explaining the reproducibility situation in more detail (https://github.com/dselsam/neurosat/wiki/On-reproducing-the-experiments). This explanation is based on an email I sent to you before you posted this issue, and probably does not contain any new information for you.
> 
> > More specifically, a couple of things that could help get me off the ground -- For the experiments described in Table 1, how many problem instances did you train on? How many epochs of training?
> 
> We performed one epoch but on the order of 10 million problems. It is performing very well way before then though, and would probably work with many passes over a smaller set of problems.
> 
> Note: you can't put 10 million files into a single directory. A crucial (omitted) optimization is fusing the SR-generation with the dimacs-to-data computations to prevent writing each individual file to the file system. This fused approach is omitted because we generated the fused data using multiple CPUs on GCE, and sent compressed versions back to a central Bucket. If I recall, the uncompressed dataset size is over 100 GB.
> 
> Note that the current unfused implementation isn't even batching at all when you sample `n` from a distribution, though this could be fixed (hackily) just by sorting the filenames in the directory to arrange by `n`.
> 
> > This doesn't look like it's converged at all
> 
> This comment doesn't make sense to me. Networks are not supposed to suffer the same cost for every single example in the dataset. In your example the network seems to be guessing `unsat` with very low confidence for every problem.

Hi, I have a question about the training epochs. You said here you only trained 1 epoch, however, in the paper, there are 26 iterations. What's the difference between iteration and epoch?

**Comment by bywbilly** [comment]
@dselsam ping!

**Comment by dselsam** [comment]
The word `epoch` is used slightly differently in different contexts, but here I just mean that we only ever trained on a given SAT problem 1 time, i.e. we effectively trained on a stream of data. The number 26 refers to how many iterations of message passing is performed when training on each individual training data point.

**Comment by bywbilly** [comment]
That is very helpful. So basically, I don't need to take care of the 26 in the training phase. I only need to set n_epoches to 1 to reproduce the results. And during the test phase, I need to set the arguments `n_rounds` to 26 to keep consistent with the paper. Do I understand correctly?

Also, here is another question, why do u need to generate two pieces of data, say grp1 and grp2?

Thanks!

@dselsam  ping

**Comment by dselsam** [comment]
> to reproduce the results

As described elsewhere, the settings in this repo are for a small, toy experiment that can run on CPU, and will not reproduce the results in the paper without more systems work.

> why do u need to generate two pieces of data, say grp1 and grp2?

This implementation doesn't stream the tensors from disk, and reads in an entire group at once into memory. So, multiple groups are needed to avoid memory explosion.

**Comment by bywbilly** [comment]
Thanks for your instant response.  

Given the fact that we cannot totally reproduce your result since you trained on 10 million data. I am wondering if we want to compare with your method tested on like 100 variables 300 clauses SAT instances, how many training samples should we use in the sake of the fairness.

Thanks! @dselsam

**Comment by dselsam** [comment]
Compare for what? The point of this project was the empirical discovery that it learns to solve the problems when trained as a classifier. There are no meaningful numbers to make go up or down. It is a total joke as an actual SAT solver. If you want to compare your other-kind-of-neural-network-guided SAT solver to a simple NeuroSAT-guided SAT solver, I suggest you compare to the SATCOMP numbers in https://arxiv.org/abs/1903.04671

**Comment by bywbilly** [comment]
Compare for the performance on solving a SAT instance using pure **deep learning models**.

**Comment by dselsam** [comment]
Hard to say without more information. Are you proposing a totally different architecture, or a different way of training a similar architecture? What kinds of problems at test time?

**Comment by bywbilly** [comment]
We just want to use NeuroSAT on larger 3-SAT instances with 100 literals and 300 clauses. We want to know how many epochs and training data do we need to train such a NeuroSAT.

**Comment by qunqunqun** [comment]
> epoch=0 | counter=4347 | n_var=28 | cost=0.785378 | is_sat=[True]
> epoch=0 | counter=4348 | n_var=30 | cost=0.613450 | is_sat=[False]
> epoch=0 | counter=4349 | n_var=33 | cost=0.612451 | is_sat=[False]
> epoch=0 | counter=4350 | n_var=21 | cost=0.609838 | is_sat=[False]
> epoch=0 | counter=4351 | n_var=18 | cost=0.784679 | is_sat=[True]
> epoch=0 | counter=4352 | n_var=22 | cost=0.783188 | is_sat=[True]
> epoch=0 | counter=4353 | n_var=32 | cost=0.782364 | is_sat=[True]
> epoch=0 | counter=4354 | n_var=39 | cost=0.608814 | is_sat=[False]
> epoch=0 | counter=4355 | n_var=19 | cost=0.607541 | is_sat=[False]
> epoch=0 | counter=4356 | n_var=32 | cost=0.608002 | is_sat=[False]

I'm trying to reproducing but I get weird result like:
epoch=0 | counter=0 | n_var=160 | cost=0.693547 | is_sat=[False, True, False, True, False, True, False, True, False, True, False, True, False, True, False, True]
epoch=0 | counter=1 | n_var=44 | cost=0.693209 | is_sat=[False, True, False, True]
epoch=0 | counter=2 | n_var=24 | cost=0.693249 | is_sat=[False, True]
epoch=0 | counter=3 | n_var=52 | cost=0.693148 | is_sat=[False, True, False, True]
epoch=0 | counter=4 | n_var=112 | cost=0.693142 | is_sat=[False, True, False, True, False, True, False, True]
.....
epoch=0 | counter=22 | n_var=210 | cost=0.693084 | is_sat=[False, True, False, True, False, True]
epoch=0 | counter=23 | n_var=144 | cost=0.693125 | is_sat=[False, True, False, True]
epoch=0 | counter=24 | n_var=222 | cost=0.693156 | is_sat=[False, True, False, True, False, True]
epoch=0 | counter=25 | n_var=228 | cost=0.693109 | is_sat=[False, True, False, True, False, True]
epoch=0 | counter=26 | n_var=312 | cost=0.693149 | is_sat=[False, True, False, True, False, True, False, True]
epoch=0 | counter=27 | n_var=240 | cost=0.693135 | is_sat=[False, True, False, True, False, True]
[0] 0.6932 (0.25, 0.25, 0.25, 0.25) [331s]

And i just change the size of problems generated, how can I get some normal results? I am depressed.. The lost is strange.
